build.sh

```diff
@@ -110,3 +113,7 @@ llvm-lto "${POSTGRES_BC}" \
 # create an archive which the Rust crate will statically link
 llvm-ar crv "${POSTGRES_PARSER_A}" "${TARGET_DIR}/raw_parser.o" || exit 1
 echo "${POSTGRES_PARSER_A};${INSTALL_DIR}"
+
+# create dynamic shared object
+CFLAGS="{$CFLAGS}" ${CC} -shared -o "${POSTGRES_PARSER_SO}" "${TARGET_DIR}/raw_parser.o" || exit 1
+echo "${POSTGRES_PARSER_SO};${INSTALL_DIR}"
```

Comment:
I thought I commented on this, but it's not here anymore...

This is going to cause `build.rs` to link against the shared library, not the static library which was being returned above on line 115.

Is this what we want?

Reply:
Ooops... my bad. Maybe should be better turn it optional creating the static by default. What’s your thoughts?

Reply:
My question is why do we need/want an .so anyways?  I don't really see how this project could benefit from it.

If you're trying to use this script for something that isn't this project, we can talk about that use case, I suppose.

Reply:
Sorry @eeeebbbbrrrr for my late answer.

You got it... I'm trying to use it for another project I'm working... the idea is create a independent shared library with Postgres parser. Similar what you did here for Rust but more generic to be used anywhere.

To simplify, my idea is join what you did here to extract the parser using LLVM to the Lukas's [libpg_query](https://github.com/lfittl/libpg_query) so we'll end up with a general purpose Postgres Parser C library.

Reply:
If that's the ultimate goal (a general purpose Postgres Parser C library), lets add a `cdylib` target to this project, and then we can either export this project's "safe" wrappers or just re-export the handful of functions from the underlying llvm static library?

We'd have to deal with the headers, but that's a thing you'd need to do anyways, and I don't think that'd be terribly difficult. In fact, we could re-generate headers from the bindgen rust bindings, paring it down to just what's necessary.

Also, you'd have the benefits of this project being easy to build thanks to cargo.

What are your thoughts?

Reply:
> If that's the ultimate goal (a general purpose Postgres Parser C library), lets add a `cdylib` target to this project, and then we can either export this project's "safe" wrappers or just re-export the handful of functions from the underlying llvm static library?
> 

Well my Rust knowledge is zero, but it sounds like a plan.

My first attempt to use it is for EnvoyProxy, because recently I helped to review a [PR to create metadata based on SQL query](https://github.com/envoyproxy/envoy/pull/11620) and the [current parser used](https://github.com/envoyproxy/sql-parser) is awful. Some minor tests I did shown that [~70% of Postgres' regression tests queries wasn't parsed](https://github.com/envoyproxy/envoy/pull/11620#issuecomment-651135825) so we need to integrate the Postgres parser to project to improve it.

Based on that I'm not sure if Envoy's maintainers will accept a library in other language different than C/C++ (but it's my own impression)

Anyway if we can create a shared library with just one header to use will be awesome, something like:

```C
#include <postgres_parser.h>
#include <stdio.h>

int main() {
  PgParserResult result;

  result = pg_parse_query("SELECT c1, c2 FROM t WHERE c1 = 999");

  printf("%s\n", result.parse_tree);

  pg_free_parse_result(result);
}
```

> We'd have to deal with the headers, but that's a thing you'd need to do anyways, and I don't think that'd be terribly difficult. In fact, we could re-generate headers from the bindgen rust bindings, paring it down to just what's necessary.
> 

Interesting... my first thought was to use the entire Postgres headers (thinking in packaging just creating dependency for -dev packages).

Reply:
> Well my Rust knowledge is zero, but it sounds like a plan.

I can help.

> Anyway if we can create a shared library with just one header to use will be awesome, something like:

Yeah, we definitely can.  It'll just be a matter of defining the API, and then tagging the functions as `extern "C"` (and I suppose #[no_mangle]).   I have to admit, I haven't really done much Rust-->C stuff (it's all been C-->Rust!, see https://github.com/zombodb/pgx), but I think it'll be straightforward now that postgres-parser takes care of all the hard stuff.

We could also expose postgres-parser's SQLScanner interface, which is great for parsing, for example, Postgres' regression test .sql files (which postgres-parser does fully parse minus a few that aren't UTF8).

> Interesting... my first thought was to use the entire Postgres headers (thinking in packaging just creating dependency for -dev packages).

I think we'd want to regenerate C-compatible headers -- Postgres' header system is a web of `#include`s and I don't think we need too much of it.  

You think the Envoy people would be interested in sponsoring this work?

Reply:
> > Anyway if we can create a shared library with just one header to use will be awesome, something like:
> 
> Yeah, we definitely can. It'll just be a matter of defining the API, and then tagging the functions as `extern "C"` (and I suppose #[no_mangle]). I have to admit, I haven't really done much Rust-->C stuff (it's all been C-->Rust!, see https://github.com/zombodb/pgx), but I think it'll be straightforward now that postgres-parser takes care of all the hard stuff.
> 

But as I said before not sure if Envoy's maintainers will accept such project (Rust instead of C/C++). Because we'll use it into Envoy and we'll need to take care of all building (bazel) and adding another dependency (Rust) may not be accepted. C/C++ library is easy to get accepted to be incorporated to project.

> We could also expose postgres-parser's SQLScanner interface, which is great for parsing, for example, Postgres' regression test .sql files (which postgres-parser does fully parse minus a few that aren't UTF8).
> 

Sure, I used it some a few months ago to create a small Postgres extension to [normalize queries](https://github.com/fabriziomello/pg_normalize_query).

> I think we'd want to regenerate C-compatible headers -- Postgres' header system is a web of `#include`s and I don't think we need too much of it.
> 

+1

> You think the Envoy people would be interested in sponsoring this work?

I don't think so... this is my company [OnGres](https://ongres.com/) initiative since we (with others) developed the https://www.envoyproxy.io/docs/envoy/v1.15.0/intro/arch_overview/other_protocols/postgres for EnvoyProxy and now we want to add more features to it like `Per query statistics` as described [here](https://github.com/envoyproxy/envoy/issues/9107#issue-526990241).

Last week was published a [first post](https://www.cncf.io/blog/2020/08/13/envoy-1-15-introduces-a-new-postgres-extension-with-monitoring-support/) about this new EnvoyProxy feature.

We're working on this project because EnvoyProxy is part of other project that we're developing called [StackGres](https://stackgres.io/).

Reply:
> we'll need to take care of all building (bazel) and adding another dependency (Rust) may not be accepted. C/C++ library is easy to get accepted to be incorporated to project.

If we went as far as to publish a library package (and a -dev package) to the various yum/apt repos, then it would just be a system dependency.  Would that work?

> Last week was published a first post about this new EnvoyProxy feature.

Yeah, I saw that on twitter.  Neat project.  A++

I have plans to build a "schema diff tool" (in Rust) that uses this project as a dependency.  I need it for `pgx` and also for supporting @ZomboDB now that it's been rewritten in Rust.  

Beyond that, I'm happy to help turn this project into something that's usable for a larger audience.  If you wanted to spec out what a C api might look like, I'll be happy to implement that interface here in Rust, and then we can figure out how to tackle the headers.  This is really one of the things Rust was meant to do...